src/CommonITILObject.php

```diff
@@ -6980,14 +6998,16 @@ public function getTimelineItems(array $options = [])
             class_exists($validation_class) && $params['with_validations']
             && ($validation_class::canView() || $params['bypass_rights'])
         ) {
-            $valitation_obj   = new $validation_class();
-            $validations = $valitation_obj->find([$foreignKey => $this->getID()]);
+            $validation_obj   = new $validation_class();
+            $validations = $validation_obj->find([$foreignKey => $this->getID()]);
             foreach ($validations as $validations_id => $validation) {
-                /** @var CommonITILValidation $valitation_obj */
-                $valitation_obj->getFromDB($validations_id);
-                $canedit = $valitation_obj->can($validations_id, UPDATE);
-                $cananswer = $valitation_obj->isCurrentUserValidationTarget(true) && ((int) $validation['status'] === CommonITILValidation::WAITING);
-                $request_key = $valitation_obj::getType() . '_' . $validations_id
+
+                /** @var CommonITILValidation $validation_obj */
+                $canedit = $validation_obj->can($validations_id, UPDATE);
+                $cananswer = ($validation_obj->canValidate($this->getID())
```

Comment:
There are other checks based on `CommonITILValidation::isCurrentUserValidationTarget()` in  `CommonITILValidation::canUpdateItem()` and  `CommonITILValidation::prepareInputForUpdate()`. I think they are obsolete and have to be updated. Maybe they could rely on `CommonITILValidation::canValidate()`, and `CommonITILValidation::isCurrentUserValidationTarget()` could probably be removed then.

Reply:
I see several things to carefully check
- CommonITILValidation::isCurrentUserValidationTarget() may or may not check the groups depending on its bool argument
- git grep shows that all calls to this method pass true as argument. It is therefore useless.
- CollonITILValidation::getTargetCriteriaForUser() also has a bool to enable groups check. No call done with groups check disabled; then this bool is also useless.

```
$ git grep "canValidate("
front/commonitilvalidation.form.php:    if ($validation->canValidate($validation->fields[$validation::$items_id])) {
src/CommonITILObject.php:                $cananswer = ($validation_obj->canValidate($this->getID())
src/CommonITILValidation.php:    public static function canValidate($items_id)
src/CommonITILValidation.php:            && !static::canValidate($item->getID())
src/Ticket.php:                  && TicketValidation::canValidate($this->fields["id"])));
tests/src/CommonITILValidation.php:        $this->boolean($validation::canValidate($itil_items_id))->isFalse();
tests/src/CommonITILValidation.php:        $this->boolean($validation::canValidate($itil_items_id))->isTrue();
tests/src/CommonITILValidation.php:        $this->boolean($validation::canValidate($itil_items_id))->isTrue();
tests/src/CommonITILValidation.php:        $this->boolean($validation::canValidate($itil_items_id))->isFalse();
tests/src/CommonITILValidation.php:        $this->boolean($validation::canValidate($itil_items_id))->isFalse();
tests/src/CommonITILValidation.php:        $this->boolean($validation::canValidate($itil_items_id))->isTrue();
tests/src/CommonITILValidation.php:        $this->boolean($validation::canValidate($itil_items_id))->isTrue();
tests/src/CommonITILValidation.php:        $this->boolean($validation::canValidate($itil_items_id))->isFalse();
```

```
$ git grep "isCurrentUserValidationTarget("
src/CommonITILValidation.php:        $is_target = $this->isCurrentUserValidationTarget(true);
src/CommonITILValidation.php:    public function isCurrentUserValidationTarget(bool $include_groups = true): bool
src/CommonITILValidation.php:        if ($this->isCurrentUserValidationTarget(true) && isset($input["status"])) {
tests/src/CommonITILValidation.php:        $this->boolean($validation->isCurrentUserValidationTarget())->isTrue();
tests/src/CommonITILValidation.php:        $this->boolean($validation->isCurrentUserValidationTarget(true))->isTrue();
tests/src/CommonITILValidation.php:        $this->boolean($validation->isCurrentUserValidationTarget(false))->isFalse();
```

In CommonITILValidation::prepareInputForUpdate() there is a logic problem before changing the call to CommonITILValidation::isCurrentUserValidationTarget(). I need to pass the linked ITIL objects ID but it may be in $this->fields[statkc::$items_id] or in $input[static::$items_id].  $forbif_fields may be empty when reaching the end od CommonITILValidation::prepareInputForUpdate(), where forbidden changes are cleaned up. 

As I don't see a use case where the FK to the ITIL object should change I strongly suggest to initialize $forbid_fields with static::$items_id. Maybe also 'entities_id'.

Let me know your thought before I add a fix.

Reply:
Indeed, I think that `static::$items_id` should not changed once item has been created, so you could move it in `$forbid_fields` initial declaration. Same for `entities_id` and `is_recursive` IMHO.

Reply:
done with 17eb48aa97

Reply:
Method removed in 6465e3295cf368b23a7f3b1aaed91506bce5280d.